L58-python-async-api.md

```diff
@@ -0,0 +1,1739 @@
+                                     response_serializer: Optional[Callable[[Any], bytes]]=None
+    ) -> grpc.aio.RpcMethodHandler:
+    """Creates an RpcMethodHandler for a stream-stream RPC method.
+
+    Args:
+      behavior: The implementation of an RPC that accepts an iterator of
+        request values and returns an iterator of response values.
+      request_deserializer: An optional behavior for request deserialization.
+      response_serializer: An optional behavior for response serialization.
+
+    Returns:
+      An RpcMethodHandler object that is typically used by grpc.Server.
+    """
+```
+
+```Python
+# grpc.aio.method_handlers_generic_handler
+def method_handlers_generic_handler(service: Text,
+                                    method_handlers: Mapping[Text, grpc.aio.RpcMethodHandler]
+    ) -> grpc.aio.GenericRpcHandler:
+    """Creates a GenericRpcHandler from RpcMethodHandlers.
+
+    Args:
+      service: The name of the service that is implemented by the
+        method_handlers.
+      method_handlers: A dictionary that maps method names to corresponding
+        RpcMethodHandler.
+
+    Returns:
+      A GenericRpcHandler. This is typically added to the grpc.Server object
+      with add_generic_rpc_handlers() before starting the server.
+    """
+```
+
+### Server Interceptor
+
+```Python
+# grpc.aio.ServerInterceptor
+class ServerInterceptor:
+    """Affords intercepting incoming RPCs on the service-side.
+
+    This is an EXPERIMENTAL API.
+    """
+
+    async def intercept_service(self,
+                          continuation: Callable[[grpc.HandlerCallDetails], grpc.aio.RpcMethodHandler],
+                          handler_call_details: grpc.HandlerCallDetails
+        ) -> grpc.aio.RpcMethodHandler:
+        """Intercepts incoming RPCs before handing them over to a handler.
+
+        Args:
+          continuation: A function that takes a HandlerCallDetails and
+            proceeds to invoke the next interceptor in the chain, if any,
+            or the RPC handler lookup logic, with the call details passed
+            as an argument, and returns an RpcMethodHandler instance if
+            the RPC is considered serviced, or None otherwise.
+          handler_call_details: A HandlerCallDetails describing the RPC.
+
+        Returns:
+          An RpcMethodHandler with which the RPC may be serviced if the
+          interceptor chooses to service this RPC, or None otherwise.
+        """
+```
+
+### Client Interceptor
+
+```Python
+# grpc.aio.UnaryUnaryClientInterceptor
+class UnaryUnaryClientInterceptor:
+    """Affords intercepting unary-unary invocations."""
+
+    async def intercept_unary_unary(self,
+                              continuation: Callable[[grpc.ClientCallDetails, Any], grpc.aio.Call[Any]],
+                              client_call_details: grpc.ClientCallDetails,
+                              request: Any
+        ) -> grpc.aio.Call[Any]:
```

Comment:
Returning an `awaitable` object from a coroutine could be problematic. Since we agreed on returning the `Call` object the  for the `UnaryUnaryMultilCallable.__call__` method, it has needed to be changed, becoming a normal function, like:

```python
    def __call__(self,
                 request,
                 timeout=None,
                 metadata=None,
                 credentials=None,
                 wait_for_ready=None,
                 compression=None):

        if timeout:
            raise NotImplementedError("TODO: timeout not implemented yet")

        if metadata:
            raise NotImplementedError("TODO: metadata not implemented yet")

        if credentials:
            raise NotImplementedError("TODO: credentials not implemented yet")

        if wait_for_ready:
            raise NotImplementedError(
                "TODO: wait_for_ready not implemented yet")

        if compression:
            raise NotImplementedError("TODO: compression not implemented yet")

        call = asyncio.ensure_future(
            self._channel.unary_unary(self._method, _common.serialize(request, self._request_serializer)))

        return Call(call, self._response_deserializer)
```

Using a coroutine, instead of a function, does not work as we would like since the awaitable object is never awaited. For example, the following line would await the coroutine but never the awaitable object

```python
response = await stub.hi(...)
```

Because the interceptors are implemented with coroutines, they would suffer from that problem where the client will be receiving as a response of the previous statement a `Call` object and never performing the proper and expected _await_..

So we would need to find a solution, either by changing how the `Call` object is being gathered or addressing the issue, which could not be fixable.

Reply:
@pfreixes I think this can be done by schedule the call to event loop directly, then wrap the returned future object as our `grpc.aio.Call` object. The implementation will be similar to:
```Python
class UnaryUnaryMultiCallable(...):

    def __call__(self, ...):
        task = asyncio.event_loop.create_task(_perform_actual_unary_unary_call())
        return grpc.aio.Call(task)
```

My biggest concern is about the performance impact, since the "create_task" may introduce more AsyncIO overhead to our framework. But turned out, the impact is not as significant as I thought.

```
lidiz-macbookpro:test-python-asyncio lidiz$ python3.7 coro_overhead.py
Spent 0.10s for 10000 runs of directly_await
10170.10 ns
Spent 0.10s for 10000 runs of schedule_task
10017.37 ns
Spent 14.91s for 1000000 runs of directly_await
14913.95 ns
Spent 15.51s for 1000000 runs of schedule_task
15512.92 ns
```

The snippet I used to run the benchmark:

```Python
import asyncio
import timeit
import time

async def work():
    return "I am not working"

async def directly_await():
    await work()

def schedule_task():
    return asyncio.create_task(work())

async def main():
    number = 10000
    start_time = time.time()
    await asyncio.gather(*(directly_await() for i in range(number)))
    elapsed_time = time.time() - start_time
    print("Spent %.2fs for %d runs of %s" % (elapsed_time, number, 'directly_await'))
    print("%.2f ns" % (elapsed_time / number * 1e9))
    
    start_time = time.time()
    await asyncio.gather(*(schedule_task() for i in range(number)))
    elapsed_time = time.time() - start_time
    print("Spent %.2fs for %d runs of %s" % (elapsed_time, number, 'schedule_task'))
    print("%.2f ns" % (elapsed_time / number * 1e9))


if __name__ == "__main__":
    asyncio.run(main())
```

Reply:
Yeps that's the plan, either by calling `ensure_future` and wrapping the output within our own `Call` object or by directly calling the `create_task`. IMO they are basically two ways of doing the same which both have little diferents.

The concern is more about where the interceptor code path should live once we have the unified call implemented, the whole problem and a likely solution has been described here [1]. TL;DR of my proposal would be having the interceptors code path within the task.

Yes, maybe its an issue for the future but seeing this as a likely blocker for the future  it's worth it to have everybody aligned since now.

[1] https://github.com/grpc/grpc/issues/20001#issuecomment-539712798